app/src/main/java/com/kickstarter/services/firebase/MessageService.kt

```diff
@@ -0,0 +1,122 @@
+package com.kickstarter.services.firebase
+
+import android.text.TextUtils
+import com.google.firebase.crashlytics.FirebaseCrashlytics
+import com.google.firebase.messaging.FirebaseMessagingService
+import com.google.firebase.messaging.RemoteMessage
+import com.google.gson.Gson
+import com.kickstarter.KSApplication
+import com.kickstarter.R
+import com.kickstarter.libs.PushNotifications
+import com.kickstarter.libs.braze.RemotePushClientType
+import com.kickstarter.models.pushdata.Activity
+import com.kickstarter.models.pushdata.GCM
+import com.kickstarter.services.ApiClientTypeV2
+import com.kickstarter.services.ApiException
+import com.kickstarter.services.apiresponses.PushNotificationEnvelope
+import com.kickstarter.services.apiresponses.PushNotificationEnvelope.Companion.builder
+import com.kickstarter.services.apiresponses.PushNotificationEnvelope.ErroredPledge
+import com.kickstarter.services.apiresponses.PushNotificationEnvelope.Survey
+import timber.log.Timber
+import javax.inject.Inject
+
+interface RefreshPushToken {
+    companion object {
+        val OK_MESSAGE = "Push Token refreshed onNewToken: "
+        val KO_MESSAGE = "Failed to register Push Token: "
+    }
+    fun invoke(apiClient: ApiClientTypeV2, newToken: String, gson: Gson, successCallback: (String) -> Unit, errorCallback: (String) -> Unit) {
+        try {
+            val response = apiClient.registerPushToken(newToken).blockingSingle()
+            val message = "$OK_MESSAGE $response"
+            successCallback.invoke(message)
+        } catch (exception: ApiException) {
+            val errorMessage = "$KO_MESSAGE ${
+            exception.errorEnvelope().errorMessage()
+            }"
+            if (exception.errorEnvelope().httpCode() != 401) // else OAuth token not authorized
+                errorCallback.invoke(errorMessage)
+        }
+    }
+}
+
+class MessageService : FirebaseMessagingService() {
+
+    @Inject
+    lateinit var pushNotifications: PushNotifications
+
+    @Inject
+    lateinit var remotePushClientType: RemotePushClientType
+
+    @Inject
+    lateinit var apiClient: ApiClientTypeV2
+
+    @Inject
+    lateinit var gson: Gson
+
+    override fun onNewToken(s: String) {
+        super.onNewToken(s)
+        val refreshHandler = object : RefreshPushToken {}
+        refreshHandler.invoke(
+            apiClient = apiClient,
+            newToken = s,
+            gson = gson,
+            successCallback = { message -> FirebaseCrashlytics.getInstance().log(message) },
+            errorCallback = { errorMessage -> FirebaseCrashlytics.getInstance().recordException(Exception(errorMessage)) }
+        )
```

Comment:
From line 59 to 65 is the newly added code, everything else if just Java-Kotlin migration